rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,517 @@
+server request with WebSocket PONG whenever the server requests it.
+
+## Wire Protocol
+
+The structure of the protocol is intentionally simple, and yet allows for
+future extension.
+
+The SDK leverages WebSocket framing, so does not need to have special fields
+for the payload size.
+
+First byte is reserved for Command codes in requests from the server, and for
+Status for SDK responses.
+
+This RFC defines the following types of the commands.
+
+| Name            | Value  | Description                                                                  |
+|-----------------|--------|------------------------------------------------------------------------------|
+| `GET_TELEMETRY` | `0x00` | Retrieve Application Telemetry from SDK in Prometheus Text Exposition Format |
+
+And the following statuses:
+
+| Name              | Value  | Description                                                |
+|-------------------|--------|------------------------------------------------------------|
+| `SUCCESS`         | `0x00` | No errors has been occured                                 |
+| `UNKNOWN_COMMAND` | `0x01` | The SDK does not support operation requested by the server |
+
+
+### `GET_TELEMETRY`
+
+The request must not have payload.
+
+The response might have payload if the SDK has metrics to report.
+
+The SDK must reset local counters after successful scheduling response for the
+network delivery.
+
+The format of the response follows the [Text-based exposition format of the
+Prometheus](https://prometheus.io/docs/instrumenting/exposition_formats/#text-based-format),
+and includes metrics and histograms collected by the SDK since the previous
+report.
+
+#### Metrics
+
+    sdk_kv_r_utimedout{agent="couchbase-net-sdk/2.4.5.0 (clr/4.0.30319.42000) (os/Microsoft Windows NT 10.0.16299.0)", id="66388CF5BFCF7522/18CC8791579B567C", node="node1", bucket="travel-sample"} 1 1695747260
+
+Where
+
+* `sdk_kv_r_utimedout` is a name of the metric, one of the:
+
+  | Metric | Description |
+  | ---- | ---- |
+  | `sdk_{service}_r_utimedout` | Unambiguous timeout |
+  | `sdk_{service}_r_atimedout` | Ambiguous timeout |
+  | `sdk_{service}_r_canceled` | Canceled |
+  | `sdk_{service}_r_total` | Total number of operations |
+
+* `{agent="couchbase-net-sdk/2.4.5.0 (clr/4.0.30319.42000) (os/Microsoft Windows NT 10.0.16299.0)", id="66388CF5BFCF7522/18CC8791579B567C", node="node1", bucket="travel-sample"}` is a list of the tags.
+
+  | Label | Description |
+  | ---- | ---- |
+  | `agent` | agent string of the SDK, as it is used everywhere (HELLO message, logs etc.) |
+  | `id` | SDK identifier string as specified by [SDK-RFC #35](0035-rto.md#client--connection-ids) in section "Client / Connection IDs" |
+  | `node` | The hostname of the node as seen in configuration, and used by the SDK to establish connection. |
+  | `alt_node` | If alternative address is used, this field should be present |
```

Comment:
Can you please include a concrete example with alternate addresses, showing which address should be used for `node` and which for `alt_node`?

Reply:
Does the server need to know the node's canonical address (the one on the default network)? If not, sending only the address used for the socket connection would simplify things on the client side, and use less bandwidth.

I confess to an ulterior motive. In Java we worked hard to isolate the complexity of alternate addresses to a single component: the cluster topology parser. Requiring other components to be aware of alternate addresses feels like a step backwards.

Reply:
the default (canonical) address would go to `node` attribute.